src/drivers/intel/dmic/dmic_nhlt.c

```diff
@@ -0,0 +1,418 @@
+{
+	int val = 0;
+	int i;
+
+	for (i = 0; i < 32; i++) {
+		val += w & 1;
+		w = w >> 1;
+	}
+
+	return val;
+}
+
+static int nhlt_dmic_dai_params_get(struct dai *dai, uint32_t *outcontrol,
+				    struct nhlt_pdm_ctrl_cfg **pdm_cfg,
+				    struct nhlt_pdm_ctrl_fir_cfg **fir_cfg)
+{
+	struct dmic_pdata *dmic = dai_get_drvdata(dai);
+	int fir_stereo = FIR_CONTROL_A_STEREO_GET(fir_cfg[0]->fir_control);
+	int mic_swap;
+
+	switch (OUTCONTROL0_OF_GET(outcontrol[dai->index])) {
+	case 0:
+	case 1:
+		dmic->dai_format = SOF_IPC_FRAME_S16_LE;
+		break;
+	case 2:
+		dmic->dai_format = SOF_IPC_FRAME_S32_LE;
+		break;
+	default:
+		dai_err(dai, "nhlt_dmic_dai_params_get(): Illegal OF bit field");
+		return -EINVAL;
+	}
+
+	switch (OUTCONTROL0_IPM_GET(outcontrol[dai->index])) {
+	case 0:
+		if (fir_stereo) {
+			dmic->dai_channels = 2;
+			dmic->enable[0] = 0x3; /* PDM0 MIC A and B */
+			dmic->enable[1] = 0x0;	/* PDM1 none */
+
+		} else {
+			dmic->dai_channels = 1;
+			mic_swap = MIC_CONTROL_PDM_CLK_EDGE_GET(pdm_cfg[0]->mic_control);
+			dmic->enable[0] = mic_swap ? 0x2 : 0x1; /* PDM0 MIC B or MIC A */
+			dmic->enable[1] = 0x0;	/* PDM1 */
+		}
+		break;
+	case 1:
+		if (fir_stereo) {
+			dmic->dai_channels = 2;
+			dmic->enable[0] = 0x0; /* PDM0 none */
+			dmic->enable[1] = 0x3;	/* PDM1 MIC A and B */
+
+		} else {
+			dmic->dai_channels = 1;
+			dmic->enable[0] = 0x0; /* PDM0 none */
+			mic_swap = MIC_CONTROL_PDM_CLK_EDGE_GET(pdm_cfg[1]->mic_control);
+			dmic->enable[1] = mic_swap ? 0x2 : 0x1; /* PDM1 MIC B or MIC A */
+		}
+		break;
+	case 2:
+		if (fir_stereo) {
+			dmic->dai_channels = 4;
+			dmic->enable[0] = 0x3; /* PDM0 MIC A and B */
+			dmic->enable[1] = 0x3;	/* PDM1 MIC A and B */
+		} else {
+			dai_info(dai, "nhlt_dmic_dai_params_get(): Illegal 4ch configuration");
+			return -EINVAL;
+		}
+		break;
+	default:
+		dai_info(dai, "nhlt_dmic_dai_params_get(): Illegal OF bit field");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int dmic_set_config_nhlt(struct dai *dai, void *spec_config)
+{
+	struct dmic_pdata *dmic = dai_get_drvdata(dai);
+	struct nhlt_pdm_ctrl_cfg *pdm_cfg[DMIC_HW_CONTROLLERS_MAX];
+	struct nhlt_pdm_ctrl_fir_cfg *fir_cfg_a[DMIC_HW_CONTROLLERS_MAX];
+	struct nhlt_pdm_ctrl_fir_cfg *fir_cfg_b[DMIC_HW_CONTROLLERS_MAX];
+	struct nhlt_pdm_fir_coeffs *fir_a[DMIC_HW_CONTROLLERS_MAX] = {NULL};
+	struct nhlt_pdm_fir_coeffs *fir_b[DMIC_HW_CONTROLLERS_MAX];
+	uint32_t out_control[DMIC_HW_FIFOS_MAX];
+	uint32_t channel_ctrl_mask;
+	uint32_t pdm_ctrl_mask;
+	uint32_t val;
+	uint32_t fir_control;
+	const uint8_t *p = spec_config;
+	int num_fifos;
+	int num_pdm;
+	int fir_length_a;
+	int fir_length_b;
+	int n;
+	int i;
+	int p_mcic = 0;
+	int p_mfira = 0;
+	int p_mfirb = 0;
+	int p_clkdiv = 0;
+	int clkdiv;
+	int comb_count;
+	int fir_decimation, fir_shift, fir_length;
+	int bf1, bf2, bf3, bf4, bf5, bf6, bf7, bf8;
+	int ret;
+
+	/* p = &nhlt_dmic_cfg[0]; */ /* Temporary blob for tests, will be replaced by IPC */
```

Comment:
time to remove this?

Reply:
Yep, and also remove the #include and the blobs header file. I just realized that I would be already already breaking bisect with this line un-commented. I need to make the topology mode default in kconfig and change later the kconfig with a PR for ipc4.